Backfill_Correction/delphiBackfillCorrection/R/main.R

```diff
@@ -0,0 +1,273 @@
+            train_data <- updated_data[[1]]
+            test_data <- updated_data[[2]]
+            sqrtscale <- updated_data[[3]]
+
+            covariates <- list(
+              Y7DAV, paste0(WEEKDAYS_ABBR, "_issue"),
+              paste0(WEEKDAYS_ABBR, "_ref"), WEEK_ISSUES, SLOPE, sqrtscale
+            )
+            params_list <- c(YITL, as.vector(unlist(covariates)))
+
+            # Model training and testing
+            prediction_results <- model_training_and_testing(
+              train_data, test_data, params$taus, params_list, params$lp_solver, 
+              params$lambda, test_lag, geo, value_type, params$cache_dir, 
+              indicator, signal, geo_level, signal_suffix,training_end_date,
+              train_models = params$train_models,
+              make_predictions = params$make_predictions
+            )
+
+            # Model objects are saved during training, so only need to export
+            # output if making predictions/corrections
+            if (params$make_predictions) {
+              test_data <- prediction_results[[1]]
+              coefs <- prediction_results[[2]]
+              test_data <- evaluate(test_data, params$taus)
+              
+              idx <- length(test_data_list[[value_type]]) + 1
+              test_data_list[[value_type]][[idx]] <- test_data
+              coef_list[[value_type]][[idx]] <- coefs
+            }
+          }# End for test lags
+        }# End for value types
+      }# End for signal suffixes
+      
+      if (params$make_predictions) {
+        for (value_type in params$value_types) {
+          for (signal_suffix in signal_suffixes) {
+            key = paste(value_type, signal_suffix)
+            test_combined <- bind_rows(test_data_list[[key]]) 
+            coef_combined <- bind_rows(coef_list[[key]]) 
+            export_test_result(test_combined, coef_combined, 
+                               indicator, signal, 
+                               geo_level, signal_suffix, lambda,
+                               training_end_date,
+                               value_type, export_dir=params$export_dir)
+          }
+        }
+      }
+      
+    }# End for geo list
+  }# End for geo type
+}
+
+#' Perform backfill correction on all desired signals and geo levels
+#' 
+#' @template params-template
+#'
+#' @importFrom dplyr bind_rows
+#' @importFrom parallel detectCores
+#' 
+#' @export
+main <- function(params) {
+  if (!params$train_models && !params$make_predictions) {
+    message("both model training and prediction generation are turned off; exiting")
+    return
+  }
+  
+  if (params$train_models) {
+    # Remove all the stored models
+    files_list <- list.files(params$cache_dir, pattern="*.model", full.names = TRUE)
+    file.remove(file.path(mydir, files_list))
+  }
+    
+  training_end_date <- as.Date(readLines(
+    file.path(params$cache_dir, "training_end_date.txt")))
```

Comment:
It's super easy to save and read the `training_end_date` this way. The only con is that we will have an extra txt file. 

We have the `training_end_date` stored as `prefix` of the saved model files as well. We can also get the date there.

Reply:
`training_end_date` will automatically change over time as we train new models. Updating params automatically in tandem with that would be finicky and require more complicated `make` logic. I think it's simpler to work off of a local cache, especially since we're already caching other objects.

We can support an optional `training_end_date` setting in `params` if we think that's important. I suppose if it were set, it would supersede the date derived from the cache.

Reply:
what does `training_end_date` get set to if there's no model yet?

Reply:
Addressed in https://github.com/cmu-delphi/covidcast-indicators/pull/1711